packages/dev/core/src/assetContainer.ts

```diff
@@ -173,29 +282,43 @@ export class AssetContainer extends AbstractScene {
             }
         };
 
-        this.transformNodes.forEach((o) => {
-            if (localOptions.predicate && !localOptions.predicate(o)) {
-                return;
-            }
+        const nodesToSort: TransformNode[] = [];
+        const idsOnSortList = new Set<number>();
 
-            if (!o.parent) {
-                const newOne = o.instantiateHierarchy(null, localOptions, (source, clone) => {
-                    onClone(source, clone);
-                });
+        for (const transformNode of this.transformNodes) {
+            if (transformNode.parent === null) {
+                this._addNodeAndDescendantsToList(nodesToSort, idsOnSortList, transformNode, localOptions.predicate);
+            }
+        }
 
-                if (newOne) {
-                    result.rootNodes.push(newOne);
-                }
+        for (const mesh of this.meshes) {
+            if (mesh.parent === null) {
+                this._addNodeAndDescendantsToList(nodesToSort, idsOnSortList, mesh, localOptions.predicate);
             }
-        });
+        }
 
-        // check if there are instanced meshes in the array, to set their new source mesh
-        const instancesExist = this.meshes.some((m) => m.getClassName() === "InstancedMesh");
-        const instanceSourceMap: TransformNode[] = [];
+        // Topologically sort nodes by parenting/instancing relationships so that all resources are in place
+        // when a given node is instantiated.
+        const sortedNodes = this._topologicalSort(nodesToSort);
 
         const onNewCreated = (source: TransformNode, clone: TransformNode) => {
             onClone(source, clone);
 
+            if (source.parent) {
+                const replicatedParentId = conversionMap[source.parent.uniqueId];
+                const replicatedParent = storeMap[replicatedParentId];
+
+                if (replicatedParent) {
+                    clone.parent = replicatedParent;
+                } else {
+                    clone.parent = source.parent;
+                }
+            }
+
+            clone.position = source.position.clone();
+            clone.rotation = source.rotation.clone();
+            clone.scaling = source.scaling.clone();

```

Comment:
Is there a need to clone here? There is a resolved comment that said we don't need, but this is still here.

Anyhow - `.copyFrom` will be more efficient than `clone()`, as those properties are already defined.

Reply:
Yup, we do need to clone on the case of instances, but I'll put the copyFrom.